File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/actions/InsertPSTCode.scala
package net.sf.latexdraw.actions

import java.text.ParseException
import org.malai.action.Action
import org.malai.undo.Undoable
import net.sf.latexdraw.badaboom.BadaboomCollector
import net.sf.latexdraw.parsers.pst.parser.PSTParser
import javax.swing.JLabel
import net.sf.latexdraw.lang.LangTool
import net.sf.latexdraw.models.interfaces.shape.IShape

/**
 * This action converts PST code into shapes and add them to the drawing.
 * <br>
 * This file is part of LaTeXDraw<br>
 * Copyright (c) 2005-2014 Arnaud BLOUIN<br>
 * <br>
 *  LaTeXDraw is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  any later version.<br>
 * <br>
 *  LaTeXDraw is distributed without any warranty; without even the
 *  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 *  PURPOSE. See the GNU General Public License for more details.<br>
 * <br>
 * 2013-02-19<br>
 * @author Arnaud Blouin
 * @since 3.0
 */
class InsertPSTCode extends Action with DrawingAction with Undoable {
	/** The code to parse. */
	var _code : Option[String] = None

	/** The status bar. */
	var _statusBar : Option[JLabel] = None

	/** The added shapes. */
	var _shapes : Option[IShape] = None



	protected def doActionBody() {
		PSTParser.cleanErrors
		try {
			new PSTParser().parsePSTCode(_code.get) match {
				case Some(group) if !group.isEmpty =>
					val sh = if(group.size()>1) group else group.getShapeAt(0)
					val br = sh.getBottomRightPoint
					val tl = sh.getTopLeftPoint
					val tx = if(tl.getX<0) -tl.getX+50 else 0
					val ty = if(tl.getY<0) -tl.getY+50 else 0

					_shapes = Some(sh)
					sh.translate(tx, ty)
					redo
					if(_statusBar.isDefined) _statusBar.get.setText(LangTool.INSTANCE.getString16("LaTeXDrawFrame.36"))
				case _ => if(_statusBar.isDefined) _statusBar.get.setText(LangTool.INSTANCE.getString16("LaTeXDrawFrame.33"))
			}
		}catch{
			case ex : Throwable =>
				BadaboomCollector.INSTANCE.add(ex)
				if(_statusBar.isDefined) _statusBar.get.setText(LangTool.INSTANCE.getString16("LaTeXDrawFrame.34"))
			}

		PSTParser.errorLogs.foreach{str => BadaboomCollector.INSTANCE.add(new ParseException(str, -1))}
		done
	}

	override def undo() {
		if(_shapes.isDefined) {
			_drawing.get.removeShape(_shapes.get)
			_drawing.get.setModified(true)
		}
	}

	override def redo() {
		if(_shapes.isDefined) {
			_drawing.get.addShape(_shapes.get)
			_drawing.get.setModified(true)
		}
	}

	def setStatusBar(value:JLabel) { _statusBar = Option(value) }

	def setCode(value:String) { _code = Option(value) }

	override def getUndoName = LangTool.INSTANCE.getStringActions("Actions.4")

	override def canDo = _code.isDefined && _drawing.isPresent

	override def hadEffect() = isDone && _shapes.isDefined

	override def isRegisterable = hadEffect()
}

File: latexdraw-core/net.sf.latexdraw/src/main/java/net/sf/latexdraw/actions/InsertPSTCode.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2018 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.actions;


import java.util.Optional;
import javafx.scene.control.Label;
import net.sf.latexdraw.badaboom.BadaboomCollector;
import net.sf.latexdraw.models.ShapeFactory;
import net.sf.latexdraw.models.interfaces.shape.IDrawing;
import net.sf.latexdraw.models.interfaces.shape.IGroup;
import net.sf.latexdraw.models.interfaces.shape.IPoint;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import net.sf.latexdraw.parsers.pst.PSTContext;
import net.sf.latexdraw.parsers.pst.PSTLatexdrawListener;
import net.sf.latexdraw.parsers.pst.PSTLexer;
import net.sf.latexdraw.parsers.pst.PSTParser;
import net.sf.latexdraw.util.LangTool;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.RecognitionException;
import org.malai.undo.Undoable;

/**
 * This action converts PST code into shapes and add them to the drawing.
 * @author Arnaud Blouin
 */
public class InsertPSTCode extends DrawingActionImpl implements Undoable {
	/** The code to parse. */
	private final String code;
	/** The status bar. */
	private final Label statusBar;
	/** The added shapes. */
	private Optional<IShape> shapes;


	public InsertPSTCode(final String codeToInsert, final Label status, final IDrawing drawingToFill) {
		super(drawingToFill);
		code = codeToInsert;
		statusBar = status;
		shapes = Optional.empty();

	}

	@Override
	protected void doActionBody() {
		try {
			final PSTLatexdrawListener listener = new PSTLatexdrawListener();
			final PSTParser parser = new PSTParser(new CommonTokenStream(new PSTLexer(CharStreams.fromString(code))));
			parser.addParseListener(listener);
			parser.pstCode(new PSTContext());

			final IGroup group = ShapeFactory.INST.createGroup();
			group.getShapes().addAll(listener.getShapes());

			if(!group.isEmpty()) {
				final IShape sh = group.size() > 1 ? group : group.getShapeAt(0);
				final IPoint tl = sh.getTopLeftPoint();
				final double tx = tl.getX() < 0d ? -tl.getX() + 50d : 0d;
				final double ty = tl.getY() < 0d ? -tl.getY() + 50d : 0d;

				shapes = Optional.of(sh);
				sh.translate(tx, ty);
				redo();

				if(statusBar != null) {
					statusBar.setText(LangTool.INSTANCE.getBundle().getString("LaTeXDrawFrame.36"));
				}
			}
		}catch(final RecognitionException ex) {
			BadaboomCollector.INSTANCE.add(ex);
			if(statusBar != null) {
				statusBar.setText(LangTool.INSTANCE.getBundle().getString("LaTeXDrawFrame.34"));
			}
		}

		done();
	}

	@Override
	public void undo() {
		shapes.ifPresent(sh -> {
			drawing.removeShape(sh);
			drawing.setModified(true);
		});
	}

	@Override
	public void redo() {
		shapes.ifPresent(sh -> {
			drawing.addShape(sh);
			drawing.setModified(true);
		});
	}

	@Override
	public String getUndoName() {
		return LangTool.INSTANCE.getBundle().getString("Actions.4");
	}

	@Override
	public boolean canDo() {
		return super.canDo() && code != null;
	}

	@Override
	public boolean hadEffect() {
		return isDone() && shapes.isPresent();
	}

	@Override
	public RegistrationPolicy getRegistrationPolicy() {
		return hadEffect() ? RegistrationPolicy.LIMITED : RegistrationPolicy.NONE;
	}
}
